js/ftx.js

```diff
@@ -471,37 +471,122 @@ module.exports = class ftx extends Exchange {
+        //                mark: "2.5587",
+        //                imfFactor: "0.0005",
+        //                lowerBound: "2.4315",
+        //                upperBound: "2.6893",
+        //                underlyingDescription: "1INCH Token",
+        //                expiryDescription: "Perpetual",
+        //                moveStart: null,
+        //                marginPrice: "2.5587",
+        //                positionLimitWeight: "20.0",
+        //                group: "perpetual",
+        //                change1h: "0.00799716356760164",
+        //                change24h: "0.004909276569004792",
+        //                changeBod: "0.008394419484511705",
+        //                volumeUsd24h: "17834492.0818",
+        //                volume: "7224898.0",
+        //                openInterest: "5597917.0",
+        //                openInterestUsd: "14323390.2279",
+        //            },
+        //            ...
+        //        ],
+        //    }
+        //
         const result = [];
         const markets = this.safeValue (response, 'result', []);
+        const allFutures = this.safeValue (allFuturesResponse, 'result');
+        const allFuturesDict = {};
+        for (let i = 0; i < allFutures.length; i++) {
+            const futureDict = allFutures[i];
+            const id = this.safeString (futureDict, 'name');
+            allFuturesDict[id] = futureDict;
+        }
         for (let i = 0; i < markets.length; i++) {
             const market = markets[i];
             const id = this.safeString (market, 'name');
+            const future = this.safeValue (allFuturesDict, id);
             const baseId = this.safeString2 (market, 'baseCurrency', 'underlying');
             const quoteId = this.safeString (market, 'quoteCurrency', 'USD');
-            const type = this.safeString (market, 'type');
             const base = this.safeCurrencyCode (baseId);
             const quote = this.safeCurrencyCode (quoteId);
+            const contract = this.safeString (market, 'type') === 'future';
+            const settleId = contract ? 'USD' : undefined;
+            const spot = !contract;
+            const margin = !contract;
+            const swap = (this.safeString (future, 'perpetual') === 'true');
+            const option = false;
+            const isFuture = contract && !swap;
+            const expiry = this.yearMonthDay (this.safeString (future, 'expiry'));
+            let type = 'spot';
+            let symbol = base + '/' + quote;
+            if (swap) {
+                type = 'swap';
+                symbol = base + '/' + quote + ':' + settleId;
+            } else if (contract) {
+                type = 'futures';
+                // TODO: delete
+                if (!expiry || !settleId || !base || !quote) {
+                    type = 'breakpoint';
+                    const y = 'stop_here';
+                    const x = 'wtf';
+                }
```

Comment:
I can't figure out this error, if I set a breakpoint inside this conditional, the breakpoint is never hit, meaning all 4 variables are defined, but I'm getting

```
Testing { exchanges: ["ftx"], symbol: "all", keys: { '--js': true, '--php': false, '--python': false, '--python-async': true, '--php-async': true, '--python-sync': true, '--php-sync': true }, maxConcurrency: 5 } (run-tests.js:255)
[100%] Testing ftx FAIL (testExchange @ run-tests.js:172)

FAILED ftx (Python 3 Async): (explain @ run-tests.js:190)

	ftx using proxy ``
	<class 'TypeError'> can only concatenate str (not "NoneType") to str 
	
	  test/test_async.py", line 558, in <module>
	    asyncio.run(main())
	
	  python3.10/asyncio/runners.py", line 44, in run
	    return loop.run_until_complete(main)
	
	  python3.10/asyncio/base_events.py", line 641, in run_until_complete
	    return future.result()
	
	  test/test_async.py", line 545, in main
	    await try_all_proxies(exchange, proxies)
	
	  test/test_async.py", line 483, in try_all_proxies
	    await load_exchange(exchange)
	
	  test/test_async.py", line 367, in load_exchange
	    await exchange.load_markets()
	
	  async_support/base/exchange.py", line 199, in load_markets
	    raise e
	
	  async_support/base/exchange.py", line 195, in load_markets
	    result = await self.markets_loading
	
	  async_support/base/exchange.py", line 185, in load_markets_helper
	    markets = await self.fetch_markets(params)
	
	  async_support/ftx.py", line 566, in fetch_markets
	    symbol = base + '/' + quote + ':' + settleId + '-' + expiry (explain @ run-tests.js:193)
FAIL ["ftx"] (run-tests.js:271)
All done, 1 failed (run-tests.js:276)
```

Reply:
@kroitor Any idea about this?

Reply:
One of `base`, `quote`, `settleId` or `expiry` is missing in one of markets?

Reply:
I tried setting a breakpoint inside the conditional but it is never hit

Reply:
In cases like these I just add `console.log (base, quote, settleId, expiry)`, run the test locally and from there it becomes obvious.